File: sidecar/src/application/lambdaService.ts
import { mkdtemp, readFile, rm, writeFile } from 'node:fs/promises';
import { tmpdir } from 'node:os';
import path from 'node:path';
import { pathToFileURL } from 'node:url';
import { config } from '../config';
import { AwsCli } from '../infrastructure/awsCli';
import { getRuntimeTemplates, type LambdaCodeInput, prepareLambdaPackage } from '../infrastructure/lambdaPackage';

export interface CreateLambdaRequest {
  functionName: string;
  runtime: string;
  handler: string;
  role?: string;
  description?: string;
  timeout?: number;
  memorySize?: number;
  code: LambdaCodeInput;
}

export interface InvokeLambdaRequest {
  payload?: unknown;
  invocationType?: 'RequestResponse' | 'Event' | 'DryRun';
}

export interface UpdateLambdaCodeRequest {
  runtime: string;
  code: LambdaCodeInput;
}

export interface UpdateLambdaConfigurationRequest {
  runtime?: string;
  handler?: string;
  role?: string;
  description?: string;
  timeout?: number;
  memorySize?: number;
  environmentVariables?: Record<string, string>;
}

export class LambdaService {
  constructor(private readonly awsCli: AwsCli) {}

  async getCapabilities() {
    return {
      endpointUrl: config.awsEndpointUrl,
      defaultRegion: config.awsRegion,
      defaultRoleArn: config.defaultLambdaRoleArn,
      runtimes: [
        { value: 'nodejs18.x', label: 'Node.js 18.x', supportsTemplate: true },
        { value: 'python3.9', label: 'Python 3.9', supportsTemplate: true },
        { value: 'go1.x', label: 'Go 1.x', supportsTemplate: false },
        { value: 'java11', label: 'Java 11', supportsTemplate: false },
        { value: 'provided.al2', label: 'Custom Runtime (provided.al2)', supportsTemplate: false },
      ],
      templates: getRuntimeTemplates(),
      sourceModes: ['template', 'inline', 'files', 'zipBase64'],
    };
  }

  async listFunctions() {
    return this.awsCli.runJson(['lambda', 'list-functions']);
  }

  async createFunction(request: CreateLambdaRequest) {
    this.assertRequired(request.functionName, 'functionName');
    this.assertRequired(request.runtime, 'runtime');
    this.assertRequired(request.handler, 'handler');

    const deploymentPackage = await prepareLambdaPackage(request.runtime, request.code);
    try {
      const args = [
        'lambda',
        'create-function',
        '--function-name',
        request.functionName,
        '--runtime',
        request.runtime,
        '--handler',
        request.handler,
        '--role',
        request.role || config.defaultLambdaRoleArn,
        '--zip-file',
        this.toAwsFileUri(deploymentPackage.zipPath, true),
      ];

      this.pushOptional(args, '--description', request.description);
      this.pushOptional(args, '--timeout', request.timeout);
      this.pushOptional(args, '--memory-size', request.memorySize);

      return await this.awsCli.runJson(args);
    } finally {
      await deploymentPackage.cleanup();
    }
  }

  async updateFunctionCode(functionName: string, request: UpdateLambdaCodeRequest) {
    this.assertRequired(functionName, 'functionName');

    const deploymentPackage = await prepareLambdaPackage(request.runtime, request.code);
    try {
      return await this.awsCli.runJson([
        'lambda',
        'update-function-code',
        '--function-name',
        functionName,
        '--zip-file',
        this.toAwsFileUri(deploymentPackage.zipPath, true),
      ]);
    } finally {
      await deploymentPackage.cleanup();
    }
  }

  async updateFunctionConfiguration(functionName: string, request: UpdateLambdaConfigurationRequest) {
    this.assertRequired(functionName, 'functionName');

    const args = ['lambda', 'update-function-configuration', '--function-name', functionName];
    this.pushOptional(args, '--runtime', request.runtime);
    this.pushOptional(args, '--handler', request.handler);
    this.pushOptional(args, '--role', request.role);
    this.pushOptional(args, '--description', request.description);
    this.pushOptional(args, '--timeout', request.timeout);
    this.pushOptional(args, '--memory-size', request.memorySize);

    if (request.environmentVariables) {
      const variables = Object.entries(request.environmentVariables)
        .map(([key, value]) => `${key}=${value}`)
        .join(',');
      args.push('--environment', `Variables={${variables}}`);
    }

    return this.awsCli.runJson(args);
  }

  async invokeFunction(functionName: string, request: InvokeLambdaRequest) {
    this.assertRequired(functionName, 'functionName');

    const workingDirectory = await mkdtemp(path.join(tmpdir(), 'floci-lambda-invoke-'));
    const payloadPath = path.join(workingDirectory, 'payload.json');
    const responsePath = path.join(workingDirectory, 'response.json');

    try {
      await writeFile(payloadPath, JSON.stringify(request.payload ?? {}), 'utf8');

      const args = [
        'lambda',
        'invoke',
        '--function-name',
        functionName,
        '--payload',
        this.toAwsFileUri(payloadPath, false),
        '--cli-binary-format',
        'raw-in-base64-out',
      ];

      this.pushOptional(args, '--invocation-type', request.invocationType);
      args.push(responsePath);

      const metadata = await this.awsCli.runJson(args);
      const rawPayload = await readFile(responsePath, 'utf8').catch(() => '');

      return {
        metadata,
        payload: this.parseJsonOrText(rawPayload),
      };
    } finally {
      await rm(workingDirectory, { recursive: true, force: true });
    }
  }

  async deleteFunction(functionName: string) {
    this.assertRequired(functionName, 'functionName');
    return this.awsCli.runJson(['lambda', 'delete-function', '--function-name', functionName]);
  }

  async getLogs(functionName: string) {
    this.assertRequired(functionName, 'functionName');
    const logGroupName = `/aws/lambda/${functionName}`;

    try {
      const streams = await this.awsCli.runJson<{ logStreams?: Array<{ logStreamName?: string }> }>([
        'logs',
        'describe-log-streams',
        '--log-group-name',
        logGroupName,
        '--order-by',
        'LastEventTime',
        '--descending',
        '--max-items',
        '5',
      ]);

      const events = [];
      for (const stream of streams.logStreams || []) {
        if (!stream.logStreamName) continue;
        const response = await this.awsCli.runJson<{ events?: unknown[] }>([
          'logs',
          'get-log-events',
          '--log-group-name',
          logGroupName,
          '--log-stream-name',
          stream.logStreamName,
          '--limit',
          '50',
        ]);
        events.push(...(response.events || []));
      }

      return { logGroupName, events };
    } catch (error) {
      return { logGroupName, events: [], warning: error instanceof Error ? error.message : String(error) };
    }
  }

  private assertRequired(value: unknown, fieldName: string) {
    if (typeof value !== 'string' || value.trim().length === 0) {
      throw new Error(`${fieldName} is required`);
    }
  }

  private pushOptional(args: string[], key: string, value: string | number | undefined) {
    if (value === undefined || value === '') return;
    args.push(key, String(value));
  }

  private toAwsFileUri(filePath: string, binary: boolean) {
    const uri = pathToFileURL(filePath).href;
    return binary ? uri.replace(/^file:\/\//, 'fileb://') : uri;
  }

  private parseJsonOrText(value: string) {
    if (!value.trim()) return null;
    try {
      return JSON.parse(value);
    } catch {
      return value;
    }
  }
}

File: mcp/floci_backend/application/lambda_service.py
import json
import os
import shutil
import tempfile
from typing import Dict, Any, List, Optional
from floci_backend.config import config
from floci_backend.infrastructure.aws_cli import AwsCli, AwsCliError
from floci_backend.infrastructure.lambda_package import prepare_lambda_package, get_runtime_templates

class LambdaService:
    def __init__(self, aws_cli: AwsCli):
        self.aws_cli = aws_cli

    async def get_capabilities(self) -> Dict[str, Any]:
        return {
            'endpointUrl': config.aws_endpoint_url,
            'defaultRegion': config.aws_region,
            'defaultRoleArn': config.default_lambda_role_arn,
            'runtimes': [
                {'value': 'nodejs18.x', 'label': 'Node.js 18.x', 'supportsTemplate': True},
                {'value': 'python3.9', 'label': 'Python 3.9', 'supportsTemplate': True},
                {'value': 'go1.x', 'label': 'Go 1.x', 'supportsTemplate': False},
                {'value': 'java11', 'label': 'Java 11', 'supportsTemplate': False},
                {'value': 'provided.al2', 'label': 'Custom Runtime (provided.al2)', 'supportsTemplate': False},
            ],
            'templates': get_runtime_templates(),
            'sourceModes': ['template', 'inline', 'files', 'zipBase64'],
        }

    async def list_functions(self) -> Dict[str, Any]:
        return await self.aws_cli.run_json(['lambda', 'list-functions'])

    async def create_function(self, request: Dict[str, Any]) -> Dict[str, Any]:
        self._assert_required(request.get('functionName'), 'functionName')
        self._assert_required(request.get('runtime'), 'runtime')
        self._assert_required(request.get('handler'), 'handler')

        deployment_package = prepare_lambda_package(request['runtime'], request.get('code', {}))
        try:
            args = [
                'lambda', 'create-function',
                '--function-name', request['functionName'],
                '--runtime', request['runtime'],
                '--handler', request['handler'],
                '--role', request.get('role', config.default_lambda_role_arn),
                '--zip-file', self._to_aws_file_uri(deployment_package.zip_path, True),
            ]
            self._push_optional(args, '--description', request.get('description'))
            self._push_optional(args, '--timeout', request.get('timeout'))
            self._push_optional(args, '--memory-size', request.get('memorySize'))

            return await self.aws_cli.run_json(args)
        finally:
            deployment_package.cleanup()

    async def update_function_code(self, function_name: str, request: Dict[str, Any]) -> Dict[str, Any]:
        self._assert_required(function_name, 'functionName')

        deployment_package = prepare_lambda_package(request.get('runtime', ''), request.get('code', {}))
        try:
            return await self.aws_cli.run_json([
                'lambda', 'update-function-code',
                '--function-name', function_name,
                '--zip-file', self._to_aws_file_uri(deployment_package.zip_path, True),
            ])
        finally:
            deployment_package.cleanup()

    async def update_function_configuration(self, function_name: str, request: Dict[str, Any]) -> Dict[str, Any]:
        self._assert_required(function_name, 'functionName')

        args = ['lambda', 'update-function-configuration', '--function-name', function_name]
        self._push_optional(args, '--runtime', request.get('runtime'))
        self._push_optional(args, '--handler', request.get('handler'))
        self._push_optional(args, '--role', request.get('role'))
        self._push_optional(args, '--description', request.get('description'))
        self._push_optional(args, '--timeout', request.get('timeout'))
        self._push_optional(args, '--memory-size', request.get('memorySize'))

        if request.get('environmentVariables'):
            variables = ','.join(f"{k}={v}" for k, v in request['environmentVariables'].items())
            args.extend(['--environment', f"Variables={{{variables}}}"])

        return await self.aws_cli.run_json(args)

    async def invoke_function(self, function_name: str, request: Dict[str, Any]) -> Dict[str, Any]:
        self._assert_required(function_name, 'functionName')

        working_directory = tempfile.mkdtemp(prefix='floci-lambda-invoke-')
        payload_path = os.path.join(working_directory, 'payload.json')
        response_path = os.path.join(working_directory, 'response.json')

        try:
            with open(payload_path, 'w', encoding='utf-8') as f:
                json.dump(request.get('payload', {}), f)

            args = [
                'lambda', 'invoke',
                '--function-name', function_name,
                '--payload', self._to_aws_file_uri(payload_path, False),
                '--cli-binary-format', 'raw-in-base64-out',
            ]
            self._push_optional(args, '--invocation-type', request.get('invocationType'))
            args.append(response_path)

            metadata = await self.aws_cli.run_json(args)

            try:
                with open(response_path, 'r', encoding='utf-8') as f:
                    raw_payload = f.read()
            except FileNotFoundError:
                raw_payload = ''

            return {
                'metadata': metadata,
                'payload': self._parse_json_or_text(raw_payload),
            }
        finally:
            shutil.rmtree(working_directory, ignore_errors=True)

    async def delete_function(self, function_name: str) -> Dict[str, Any]:
        self._assert_required(function_name, 'functionName')
        return await self.aws_cli.run_json(['lambda', 'delete-function', '--function-name', function_name])

    async def get_logs(self, function_name: str) -> Dict[str, Any]:
        self._assert_required(function_name, 'functionName')
        log_group_name = f"/aws/lambda/{function_name}"

        try:
            streams = await self.aws_cli.run_json([
                'logs', 'describe-log-streams',
                '--log-group-name', log_group_name,
                '--order-by', 'LastEventTime',
                '--descending',
                '--max-items', '5',
            ])

            events = []
            for stream in streams.get('logStreams', []):
                if not stream.get('logStreamName'):
                    continue
                response = await self.aws_cli.run_json([
                    'logs', 'get-log-events',
                    '--log-group-name', log_group_name,
                    '--log-stream-name', stream['logStreamName'],
                    '--limit', '50',
                ])
                events.extend(response.get('events', []))

            return {'logGroupName': log_group_name, 'events': events}
        except Exception as error:
            message = str(error)
            if isinstance(error, AwsCliError):
                message = error.stderr.strip() or str(error)
            return {'logGroupName': log_group_name, 'events': [], 'warning': message}

    def _assert_required(self, value: Any, field_name: str) -> None:
        if not isinstance(value, str) or not value.strip():
            raise ValueError(f"{field_name} is required")

    def _push_optional(self, args: List[str], key: str, value: Any) -> None:
        if value is None or value == '':
            return
        args.extend([key, str(value)])

    def _to_aws_file_uri(self, file_path: str, binary: bool) -> str:
        # Create file URI similar to pathToFileURL in node
        abs_path = os.path.abspath(file_path).replace('\\', '/')
        if not abs_path.startswith('/'):
            abs_path = '/' + abs_path
        uri = f"file://{abs_path}"
        return uri.replace('file://', 'fileb://') if binary else uri

    def _parse_json_or_text(self, value: str) -> Any:
        if not value.strip():
            return None
        try:
            return json.loads(value)
        except json.JSONDecodeError:
            return value
